staging/src/k8s.io/apiserver/pkg/authentication/instrumentedauthenticator/metrics_test.go

```diff
@@ -0,0 +1,327 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package instrumentedauthenticator
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"strings"
+	"testing"
+
+	"k8s.io/apiserver/pkg/authentication/authenticator"
+	"k8s.io/apiserver/pkg/authentication/request/union"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/component-base/metrics/testutil"
+)
+
+func TestMetrics(t *testing.T) {
+	// Excluding authentication_duration_seconds since it is difficult to predict its values.
+	metrics := []string{
+		"authenticated_user_requests",
+		"authentication_attempts",
+	}
+
+	testCases := []struct {
+		desc        string
+		want        string
+		request     authenticator.Request
+		apiAudience authenticator.Audiences
+	}{
+		{
+			desc: "auth ok",
+			request: WrapRequest(
+				authenticator.RequestFunc(
+					func(_ *http.Request) (*authenticator.Response, bool, error) {
+						return &authenticator.Response{User: &user.DefaultInfo{Name: "admin"}}, true, nil
+					}),
+				"instrumented-auth"),
+			want: `
+				# HELP authenticated_user_requests [ALPHA] Counter of authenticated requests broken out by username.
+				# TYPE authenticated_user_requests counter
+				authenticated_user_requests{authenticator="instrumented-auth", username="admin"} 1
+		    # HELP authentication_attempts [ALPHA] Counter of authenticated attempts.
+		    # TYPE authentication_attempts counter
+		    authentication_attempts{authenticator="instrumented-auth", result="success"} 1
+				`,
+		},
+		{
+			desc: "auth failed with error",
+			request: WrapRequest(
+				authenticator.RequestFunc(
+					func(_ *http.Request) (*authenticator.Response, bool, error) {
+						return nil, false, errors.New("some error")
+					}),
+				"instrumented-auth"),
+			want: `
+		    # HELP authentication_attempts [ALPHA] Counter of authenticated attempts.
+		    # TYPE authentication_attempts counter
+		    authentication_attempts{authenticator="instrumented-auth", result="error",} 1
```

Comment:
I see your point.

How would we define an alert based on the error rate per authenticator? 
It does not seem to be possible to do so since we can't differentiate these two scenarios:
1. webhook auth is having a problem
2. the token was intended to another token authenticator higher up the stack

Reply:
I think we might have to make this stateful by keeping track of all metrics that we may want to assert.  On success, we would only report the last one (i.e. the successful authentication with authenticator named `foo`).  On failure, we would report failure metrics for all authenticators where `err != nil`.  On the `nil, false, nil` case, we report no metrics related to authenticator name.

WDYT?

Reply:
I believe there five use-cases:
1. An authenticator receives an incompatible credential. In this case, such an authenticator should return `nil, false, nil` and such authentication attempts should be ignored in the `recordMetrics` since this is expected.
2. An authenticator receives a compatible credential, but the authentication fails (ex. incorrect password). In this case, `nil, false, nil` is **also** returned. This case is interesting from the monitoring point of view since it may indicate a security incident - somebody is trying to brute-force a password, and I think we should be recording this.
3. The supplied credential is compatible with the authenticator, but an error occurred during validation (ex. network failure), in this case we return, nil, false, someError`. We certainly, want to record this fact.
4. Happy path - compatible credentials and successful authentication.
5. None of the authenticators were able to authenticate a request/token.

Therefore
1. As you suggested, to distinguish between 1 and 2 we need to know the final result produced by the union authenticator. If the final outcome is a success, we could ignore case 2. 
2. For use-case 5, I think we need to instrument our union authenticators - they could have their own instrumentation to track requests that failed the chain of authenticators.
3. The current state of this PR properly handles cases 3 and 4.

Actually, by instrumenting union authenticators (to emit a metric when none of the authenticators in the chain were successful) we can safely ignore the issue of not being able to distinguish cases 1 and 2. In other words, if there a brute-force attack in progress, this would be captured by the union authenticator. 

> On the nil, false, nil case, we report no metrics related to authenticator name.
   I believe this is already what this PR does.
```
if resp == nil && err == nil {
  // When both of these are nil, we assume that the request was incompatible with the authenticator.
  return
}
```

In summary, I think the current logic in this PR is correct, we just need to instrument the union authenticators. WDYT?